File: frontend/Concord/src/api/serversService.js
import { request } from './httpClient';
import { buildPagingQuery } from './pagination';

export async function createServer({ Name, IconUrl = null }) {
  return request('Servers', { method: 'POST', body: { Name, IconUrl } });
}

export async function getMyServers() {
  return request('Servers');
}

export async function joinServer(inviteCode) {
  return request(`Servers/Join/${inviteCode}`, { method: 'POST' });
}

export async function updateServer(serverId, { Name, IconUrl = null }) {
  return request(`Servers/${serverId}`, { method: 'PUT', body: { Name, IconUrl } });
}

export async function leaveServer(serverId) {
  await request(`Servers/${serverId}/Members/Me`, { method: 'DELETE' });
}

export async function removeServerMember(serverId, userId) {
  await request(`Servers/${serverId}/Members/${userId}`, { method: 'DELETE' });
}

export async function deleteServer(serverId) {
  await request(`Servers/${serverId}`, { method: 'DELETE' });
}

export async function transferServerOwnership(serverId, newOwnerUserId) {
  await request(`Servers/${serverId}/Ownership/Transfer`, {
    method: 'POST',
    body: { NewOwnerUserId: newOwnerUserId },
  });
}

export async function getServerMembers(serverId, { page, pageSize } = {}) {
  return request(`Servers/${serverId}/Members`, { query: buildPagingQuery(page, pageSize) });
}

export async function createServerInvite(serverId, { ExpiresAtUtc = null, MaxUses = null } = {}) {
  return request(`Servers/${serverId}/Invites`, { method: 'POST', body: { ExpiresAtUtc, MaxUses } });
}

export async function listServerInvites(serverId) {
  return request(`Servers/${serverId}/Invites`);
}

File: mobile/concord/lib/api/servers_service.dart
import 'api_client.dart';
import 'models/invite_response.dart';
import 'models/paged_result.dart';
import 'models/server_member_summary.dart';
import 'models/server_response.dart';

class ServersService {
  ServersService(this._client);

  final ApiClient _client;

  Future<ServerResponse> createServer({required String name, String? iconUrl}) async {
    final data = await _client.post('/Servers', body: {'Name': name, 'IconUrl': iconUrl});
    return ServerResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<List<ServerResponse>> listServers() async {
    final data = await _client.get('/Servers');
    return (data as List<dynamic>)
        .map((e) => ServerResponse.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<ServerResponse> joinServer(String inviteCode) async {
    final data = await _client.post('/Servers/Join/$inviteCode');
    return ServerResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<void> leaveServer(String serverId) => _client.delete('/Servers/$serverId/Members/Me');

  Future<void> kickMember(String serverId, String userId) =>
      _client.delete('/Servers/$serverId/Members/$userId');

  Future<void> transferOwnership({required String serverId, required String newOwnerUserId}) {
    return _client.post(
      '/Servers/$serverId/Ownership/Transfer',
      body: {'NewOwnerUserId': newOwnerUserId},
    );
  }

  Future<PagedResult<ServerMemberSummary>> listMembers(
    String serverId, {
    int page = 1,
    int pageSize = 30,
  }) async {
    final data = await _client.get(
      '/Servers/$serverId/Members',
      query: {'page': page, 'pageSize': pageSize},
    );
    return PagedResult.fromJson(data as Map<String, dynamic>, ServerMemberSummary.fromJson);
  }

  Future<InviteResponse> createInvite(
    String serverId, {
    DateTime? expiresAtUtc,
    int? maxUses,
  }) async {
    final data = await _client.post(
      '/Servers/$serverId/Invites',
      body: {'ExpiresAtUtc': expiresAtUtc?.toIso8601String(), 'MaxUses': maxUses},
    );
    return InviteResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<List<InviteResponse>> listInvites(String serverId) async {
    final data = await _client.get('/Servers/$serverId/Invites');
    return (data as List<dynamic>).map((e) => InviteResponse.fromJson(e as Map<String, dynamic>)).toList();
  }
}
